tests/suppressionlist_test.php

```diff
@@ -0,0 +1,174 @@
+     *
+     * This test covers the basic CRUD (Create, Read, Update, Delete) operations
+     * for the suppression list database table.
+     *
+     */
+    public function test_suppression_list_crud() {
+        global $DB;
+
+        $this->resetAfterTest(true);
+
+        // Create some test data.
+        $testdata = [
+            ['email' => 'test1@example.com', 'reason' => 'Bounce', 'created_at' => '2024-03-01 10:00:00'],
+            ['email' => 'test2@example.com', 'reason' => 'Complaint', 'created_at' => '2024-03-02 11:00:00'],
+        ];
+
+        // Insert test data.
+        foreach ($testdata as $data) {
+            $record = new stdClass();
+            $record->email = $data['email'];
+            $record->reason = $data['reason'];
+            $record->created_at = $data['created_at'];
+            $record->timecreated = time();
+            $record->timemodified = time();
+
+            $DB->insert_record('tool_emailutils_suppression', $record);
+        }
+
+        // Test retrieval.
+        $records = $DB->get_records('tool_emailutils_suppression');
+        $this->assertCount(2, $records);
+
+        // Test specific record retrieval.
+        $record = $DB->get_record('tool_emailutils_suppression', ['email' => 'test1@example.com']);
+        $this->assertEquals('Bounce', $record->reason);
+
+        // Test update.
+        $record->reason = 'Updated Reason';
+        $DB->update_record('tool_emailutils_suppression', $record);
+        $updatedrecord = $DB->get_record('tool_emailutils_suppression', ['email' => 'test1@example.com']);
+        $this->assertEquals('Updated Reason', $updatedrecord->reason);
+
+        // Test delete.
+        $DB->delete_records('tool_emailutils_suppression', ['email' => 'test2@example.com']);
+        $remainingrecords = $DB->get_records('tool_emailutils_suppression');
+        $this->assertCount(1, $remainingrecords);
+    }
+
+    /**
+     * Test the scheduled task for updating the suppression list.
+     *
+     * This test creates a mock task that overrides the AWS SES client
+     * to return predefined test data. It then executes the task and
+     * verifies that the suppression list in the database is updated correctly.
+     *
+     * @return void
+     * @covers \tool_emailutils\task\update_suppression_list::execute
+     * @covers \tool_emailutils\task\update_suppression_list::fetch_aws_ses_suppression_list
+     * @throws dml_exception
+     */
+    public function test_update_suppression_list_task(): void {
+        global $DB;
+        $this->resetAfterTest(true);
+
+        // Set up mock AWS credentials (these won't be used, but let's set them anyway).
+        set_config('aws_region', 'eu-west-2', 'tool_emailutils');
+        set_config('aws_key', 'testkey', 'tool_emailutils');
+        set_config('aws_secret', 'testsecret', 'tool_emailutils');
+
+        $mocktask = new class extends \tool_emailutils\task\update_suppression_list {
```

Comment:
This class overrides all of the real logic, such that this test is only actually running mocked code. As a demonstration, if you add `throw new Exception('Foo');` at the top of the real `execute` or `fetch_aws_ses_suppression_list` methods, this test still passes because that code is never being run.

Ideally what you want is a way to override the `$sesv2` client object with a mocked version that returns a known set of values, then you can test that the rest of the code behaves as expected. In Moodle 4.4 we have dependency injection to help with this, but since this needs to work on 4.3, another approach would be to define the client in a property of the class. In the real code, if this is not set it can create an instance of `\Aws\SesV2\SesV2Client`, but in the test you can set it to a mocked version before you call `execute()`.